streamline the database interfaces and add pgx adapter

julesjcraske opened the pull request:
I am opening this draft PR to see if you are interested in this direction for the library (trimming down the interfaces to only what is used). If so I can put the effort in to get it up to your required standards and resolve the final issues. If not I can close it and keep my fork.

Thanks for all the great work you have done on the lib so far!

I want to use watermill to fulfil the outbox pattern (persisting events that will be asynchronously broadcast as part of the main persistence transaction) but my company mainly uses the pgx library.

This change makes this library more flexible to allow for the use of the pgx library. I can keep the pgx adapter itself elsewhere, outside of this repo if you would like.

- Strips out all of the unused methods declared in the database Beginner and Executor (totally unused) and ContextExecutor interfaces.
- Adds interfaces for the Exec Result type and the Query Rows type.
- Adds small plumbing adapters for the StdSQL and Pgx types

m110 commented on pkg/sql/sql_adapter.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+// BeginTx converts the stdSQL.Tx struct to our Tx interface
+func (c *StdSQLBeginner) BeginTx(ctx context.Context, options *sql.TxOptions) (Tx, error) {
```
How about making these value receivers instead of pointers? There should be no change performance-wise since the embedded values are pointers anyway, and it would make for a bit simpler construction of these. :)

m110 commented:
Thanks for contributing @julesjcraske and I'm sorry it took us so long to reply!

This looks great, and I since we're getting more people asking about using specific database frameworks with the outbox pattern, I'm happy to go ahead with it. Since you changed the public interfaces anyway, I think we can't keep backward compatibility and will need to bump major version, unless you see some way?

julesjcraske commented:
>  I think we can't keep backward compatibility and will need to bump major version, unless you see some way?
Thanks for taking a look, its great to here you are interested in the concept. I will try and find time to explore the options backwards compat and finishing touches on the PR over the next few weeks.

julesjcraske commented on pkg/sql/pgx/pgx_adapter.go:
```diff
@@ -0,0 +1,128 @@
+	return Rows{rows}, err
+}
+
+func (t Tx) Rollback() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	return t.Tx.Rollback(ctx)
+}
+
+func (t Tx) Commit() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	return t.Tx.Commit(ctx)
+}
```
Because the stdSQL interface this interface uses doesn't take a context, it looks like I decided to create a context and add a timeout here. 

I don't want to refactor too much to make context available here in this PR.

1. I could just use a background context with no timeout
2. I could make the timeout configurable when creating this adapter
3. I could keep a hardcoded default timeout as it is currently (I don't like this as much)

julesjcraske commented:
Update: I have created a pull request on the pgxtype repository to add the xid8 type that is currently used in the postgres schema. https://github.com/jackc/pgtype/pull/220

There are still some more fixes required and issues to resolve after that, but that gets the majority of the tests passing with the new adapters.

m110 commented:
Hey @julesjcraske, I see the xid8 is supported natively in pgx v5. Should we stick to this version then?

davideimola commented:
Hey @julesjcraske and @m110,

I have tried to create a fork for pgx and upgrade everything to v5 which works just fine.

If you want to take a look I put everything on this repo.

https://github.com/davideimola/watermill-pgx

julesjcraske commented:
Upgrading pgx to v5 in this PR was easy actually thanks.

The problem I have had is that with pgx even with the same DefaultPostgresqlSchema, the subscriber initialising queries that create tables have to be run inside a transaction otherwise I get the errors: 
```
    test_pubsub.go:347: could not initialize schema: ERROR: duplicate key value violates unique constraint "pg_class_relname_nsp_index" (SQLSTATE 23505)
    test_pubsub.go:347: could not initialize schema: ERROR: duplicate key value violates unique constraint "pg_class_relname_nsp_index" (SQLSTATE 23505)
error on publish: could not insert message as row: ERROR: relation "test_pgx_topic_83ad5299-10bf-406f-bf15-87c3161443be_11" does not exist (SQLSTATE 42P01), 4 retries left
error on publish: could not insert message as row: ERROR: relation "test_pgx_topic_83ad5299-10bf-406f-bf15-87c3161443be_7" does not exist (SQLSTATE 42P01), 4 retries left
```

I added a mechanism to run schema transactions inside a transaction and the tests then run. There might be another way to solve it? I followed advice in https://stackoverflow.com/questions/74261789/postgres-create-table-if-not-exists-%E2%87%92-23505

But now the schemas initialise, when querying for messages I get the error logs:
```
[watermill] 2024/10/16 09:50:53.895518 subscriber.go:280: 	level=ERROR msg="could not commit tx for querying message" consumer_group=test err="commit unexpectedly resulted in rollback" subscriber_id=01JAA8BPYF5TBCJ0ZJBQNBMSHE topic=topic_f8ed06a2-10d3-4f92-bbb7-062b0f70e5c2 
```

These see, rather dangerous as the tests pass, but errors are only logged. Shouldn't the transaction commit be outside the defer?

m110 commented:
Thanks for looking into it @julesjcraske!

Wouldn't this indicate that two subscribers, in parallel, try to create a table with the same name? 🤔 I guess this could happen if tests initialize the schema with both the publisher and subscriber. Perhaps we could adjust how they do it instead.

julesjcraske commented:
> Thanks for looking into it @julesjcraske!
> 
> Wouldn't this indicate that two subscribers, in parallel, try to create a table with the same name? 🤔 I guess this could happen if tests initialize the schema with both the publisher and subscriber. Perhaps we could adjust how they do it instead.

The weird thing is, the Postgres with standard SQL adapter tests work fine, so it doesn't seem to actually be a database issue but an issue with the library used to connect. It definitely warrants more investigation.

julesjcraske commented:
Sure there isn't much left to do, I think there are a few failing tests and
there is the weird issue I had with needing to create tables inside a
transaction with a lock that I wasn't sure if adding the feature was
necessary or if something else was actually wrong.

On Tue, 12 Nov 2024, 10:39 Davide Imola, ***@***.***> wrote:

> Hi @m110 <https://github.com/m110> and @julesjcraske
> <https://github.com/julesjcraske>!
> Is there any way that I could help with this PR? I needed to solve my
> compatibility issues as I am actually running a custom implementation
> instead of the actual library!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ThreeDotsLabs/watermill-sql/pull/29#issuecomment-2470181161>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABONZ7CT353KX42QYF6VRX32AHLFJAVCNFSM6AAAAABNAQZVISVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINZQGE4DCMJWGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

m110 commented:
Hey @julesjcraske. I looked into the logged errors, and as far as I understand, they could also be happening before your changes, as we wouldn't usually log them if the tests pass correctly. 🤔 So perhaps this is unrelated?

julesjcraske commented:
> Hey @julesjcraske. I looked into the logged errors, and as far as I understand, they could also be happening before your changes, as we wouldn't usually log them if the tests pass correctly. 🤔 So perhaps this is unrelated?

Yeah i noticed the logs are don't actually cause a test failure.

You are right I did seem to get all the tests passing by adding the transaction around the schema migrations.

I have just added a small improvement to duplicated code, and added a backward compatibility fix for NewSubscriber, introducing NewSubscriberV2 with new interface. I could also make NewSubscriber accept a new interface value as well, but that would encourage its usage. Alternatively you might prefer to make a backward incompatible change, but I don't think it it worth if it it is this easily avoidable?

m110 commented:
Hey @julesjcraske, we're at 1.4 RC1 at the moment, so perhaps it's fine to change the API. Since the interfaces change, we might as well do it everywhere.

julesjcraske commented:
I have resolved conflicts with the main branch and removed the backwards compatibility suggestion.

rhhammond commented:
Hi, is there an ETA for when this review will be completed?

m110 commented:
Hey @rhhammond, it's very close — I'm preparing a PR with suggestions now. :)

m110 commented:
Hey @julesjcraske, GitHub doesn't make it easy to create a PR to a forked repository, so I just pushed my commit here 😅 https://github.com/ThreeDotsLabs/watermill-sql/commit/08c87fbcb3de935507e392e7deb97c823466bb0e

Suggestions:
* Moved the adapters to the main `sql` package to stick to the existing convention.
* Added constructors for creating the beginners.
* Make the pgx Tx use the context passed to the beginner.

Let me know your thoughts. If you're happy with it, I think we're ready to merge!

m110 commented:
Two more things on my todo list:
* The tests are broken, looking into it.
* Perhaps we could create some easier to use API for the advisory lock, so that the user doesn't need to come up with a number (e.g., provide a bool setting and a default value, so it's easier to use).

m110 commented:
@julesjcraske After discussing this with @roblaszczak, we thought it might be safer to make the init in the transaction by default. Otherwise, it may not be clear it needs to be enabled. Please check the changes and let me know what you think!

julesjcraske commented:
> @julesjcraske After discussing this with @roblaszczak, we thought it might be safer to make the init in the transaction by default. Otherwise, it may not be clear it needs to be enabled. Please check the changes and let me know what you think!

Those all seem like good changes 👍

m110 commented:
Thank you for contributing this @julesjcraske! 🥇 
I've released `v4.0.0-rc.2` so we can test this in examples first.
I'll follow up with migration instructions as well. :)

Amf1k commented:
@julesjcraske @m110 Hi, I noticed one problem in v4.0.0-rc2.
I can't reuse pgx transaction (which is already open) to save something to db and put message to outbox table.

```
	err := pgx.BeginTxFunc(ctx, m.db, pgx.TxOptions{}, func(tx pgx.Tx) error {
		_, err := tx.Exec(ctx, "INSERT INTO table (id) VALUES ($1)", 123)
		if err != nil {
			return fmt.Errorf("failed to save message: %w", err)
		}

		var publisher message.Publisher

		publisher, err = sql.NewPublisher(sql.BeginnerFromPgx(tx), sql.PublisherConfig{ <<--- We have a problem
			SchemaAdapter: sql.DefaultPostgreSQLSchema{},
		}, watermill.NopLogger{})

		if err != nil {
			return err
		}

		publisher = forwarder.NewPublisher(publisher, forwarder.PublisherConfig{
			ForwarderTopic: "test",
		})

		err = publisher.Publish("test", message.NewMessage(watermill.NewUUID(), []byte("test")))
		if err != nil {
			return err
		}

		return nil
	})
```

Cause pgx.Tx interface don't have `BeginTx(ctx context.Context, options pgx.TxOptions) (pgx.Tx, error)`

m110 commented:
Good point, thank you @Amf1k. I think we need to provide separate `BeginnedFromPgx` and `ContextExecutorFromPgx` wrappers. Also, probably a separate one for pgxConn and pgxTx.

m110 commented:
@Amf1k I had this `TxFromPgx` method in the PR but then decided to drop it. It seems it's useful after all :) https://github.com/ThreeDotsLabs/watermill-sql/pull/42

Abdullah-AlAttar commented:
I need this feature so I am playing around with it. but I have faced this issue mentioned [here](https://github.com/jackc/pgx/discussions/2137 ) regarding scanning xid8 type

rhhammond commented:
@m110 

> @Amf1k I had this TxFromPgx method in the PR but then decided to drop it. It seems it's useful after all :) https://github.com/ThreeDotsLabs/watermill-sql/pull/42

Does it matter that there is not a way to supply `ctx`  with `TxFromPgx`?  It looks like [PgxTx.Rollback()](https://github.com/ThreeDotsLabs/watermill-sql/blob/master/pkg/sql/adapters_pgx.go#L93) will panic if `ctx` is nil.
